server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -71,41 +71,55 @@
+    // need to synchronize access to these variables.
     protected Manifest previousManifest;
     protected ClusterState previousClusterState;
     protected boolean incrementalWrite;
 
-    public GatewayMetaState(Settings settings, MetaStateService metaStateService,
-                            MetaDataIndexUpgradeService metaDataIndexUpgradeService, MetaDataUpgrader metaDataUpgrader,
-                            TransportService transportService, ClusterService clusterService) throws IOException {
+    public GatewayMetaState(Settings settings, MetaStateService metaStateService) {
         this.settings = settings;
         this.metaStateService = metaStateService;
-        this.transportService = transportService;
-        this.clusterService = clusterService;
+    }
 
-        upgradeMetaData(metaDataIndexUpgradeService, metaDataUpgrader);
-        initializeClusterState(ClusterName.CLUSTER_NAME_SETTING.get(settings));
+    public void start(TransportService transportService, ClusterService clusterService,
+                      MetaDataIndexUpgradeService metaDataIndexUpgradeService, MetaDataUpgrader metaDataUpgrader) {
+        assert previousClusterState == null : "should only start once, but already have " + previousClusterState;
+        try {
+            upgradeMetaData(metaDataIndexUpgradeService, metaDataUpgrader);
+            initializeClusterState(ClusterName.CLUSTER_NAME_SETTING.get(settings));
+        } catch (IOException e) {
+            throw new ElasticsearchException("failed to load metadata", e);
+        }
         incrementalWrite = false;
-    }
 
-    public PersistedState getPersistedState(Settings settings, ClusterApplierService clusterApplierService) {
-        applyClusterStateUpdaters();
-        if (DiscoveryNode.isMasterNode(settings) == false) {
-            // use Zen1 way of writing cluster state for non-master-eligible nodes
-            // this avoids concurrent manipulating of IndexMetadata with IndicesStore
-            clusterApplierService.addLowPriorityApplier(this);
-            return new InMemoryPersistedState(getCurrentTerm(), getLastAcceptedState());
+        applyClusterStateUpdaters(transportService, clusterService);
+        if (DiscoveryModule.DISCOVERY_TYPE_SETTING.get(settings).equals(DiscoveryModule.ZEN_DISCOVERY_TYPE)) {
+            // only for tests that simulate a mixed Zen1/Zen2 clusters, see Zen1IT
+            if (isMasterOrDataNode()) {
+                clusterService.addLowPriorityApplier(this::applyClusterState);
+            }
+            persistedState.set(new InMemoryPersistedState(getCurrentTerm(), getLastAcceptedState()));
+        } else {
+            if (DiscoveryNode.isMasterNode(settings) == false) {
+                if (DiscoveryNode.isDataNode(settings)) {
+                    // use Zen1 way of writing cluster state for non-master-eligible nodes
```

Comment:
I would appreciate if you add an extended comment what is "Zen1" style of writing cluster state and is "Zen2" style of writing cluster.

Reply:
Does this clarify? https://github.com/elastic/elasticsearch/pull/46326/files#diff-e5547fdd4d81fb437b07570d8c0de04fR118-R131

Reply:
The explanation looks very good, thank you!